keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,767 @@
+solution in such a way to allow future extension to facilitate the cloud administratore to disable the feature and introduce a custom external agents to fill the NodeResourceTopology data.
+
+### Changes to the API
+
+Code responsible for working with NodeResourceTopology API will be placed in the staging directory
+at path staging/src/k8s.io/api/node/v1/types.go.
+
+```go
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopology is a specification for a hardware resources
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	TopologyPolicies []string 		   `json:"topologyPolicies"`
+	Cells            map[string]Cell   `json:"cells"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopologyList is a list of NodeResourceTopology resources
+type NodeResourceTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []NodeResourceTopology `json:"items"`
+}
+
+// Cell is the spec for a NodeResourceTopology resource
+type Cell struct {
+    Name       string           `json:"name"`
+    Type       string           `json:"type"`
+    Parent     string           `json:"parent,omitempty"`
+    Costs      []CostInfo       `json:"costs,omitempty"`
+    Attributes []AttributeInfo  `json:"attributes,omitempty"`
+    Resources  []ResourceInfo   `json:"resources,omitempty"`
+}
+
+type ResourceInfo struct {
+    Name        string `json:"name"`
+    Allocatable string `json:"allocatable"`
+    Capacity    string `json:"capacity"`
+}
+
+type CostInfo struct {
+    Name  string `json:"name"`
+    Value int	 `json:"value"`
+}
+
+type AttributeInfo struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+}
+// Kubelet writes to NodeResourceTopology
+// and scheduler plugin reads from it 
+// Real world example of how these fields are populated is as follows:
+// Cells:
+//   Name:     node-1
+//   Type:     Node
+//   Costs:
+//     Name:   node-0
+//     Value:  20
+//     Name:   node-1
+//     Value:  10
+//   Attributes:
+//    Name: performance-profile
+//    Value: high-performance-profile
+//   Resources:
+//     Name:         example.com/deviceB
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         example.com/deviceA
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         cpu
+//     Allocatable:  4
+//     Capacity:     4
+
+```
+
+
+### Plugin implementation details
+
+### Description of the Algorithm
+
+The algorithm which of the scheduler plugin is as follows:
+
+1. At the filter extension point of the plugin, the QoS class of the pod is determined, in case it is a best effort pod or the 
+   Topology Manager Policy configured on the node is not single-numa-node, the node is not considered for scheduling
```

Comment:
if the node is not configured with single-numa-node, shouldn't this node pass the filter?

Reply:
What happens if the `NodeResourceTopology` is not present?